runners/core-java/src/main/java/org/apache/beam/runners/core/ReduceFnRunner.java

```diff
@@ -919,7 +919,7 @@ private boolean needToEmit(boolean isEmpty, boolean isFinished, PaneInfo.Timing
       // The pane has elements.
       return true;
     }
-    if (timing == Timing.ON_TIME) {
+    if (!isFinished && timing == Timing.ON_TIME) {
```

Comment:
I think this might be wrong. The `isFinished` boolean I think actually indicates whether it is currently in the process of finishing. If it is actually _already_ finished, then I think this is unreachable. The logic is quite tangled, so I could be wrong. Can you describe what brought you to this change?